packages/datetime/src/dateTimePicker.tsx

```diff
@@ -60,46 +62,52 @@ export class DateTimePicker extends AbstractComponent<IDateTimePickerProps, IDat
     public constructor(props?: IDateTimePickerProps, context?: any) {
         super(props, context);
 
+        const initialValue = (this.props.value != null) ? this.props.value : this.props.defaultValue;
         this.state = {
-            value: (this.props.value != null) ? this.props.value : this.props.defaultValue,
+            dateValue: initialValue,
+            timeValue: initialValue,
         };
     }
 
     public render() {
+        const value = DateUtils.getDateTime(this.state.dateValue, this.state.timeValue);
         return (
             <div className={classNames(Classes.DATETIMEPICKER, this.props.className)}>
                 <DatePicker
                     {...this.props.datePickerProps}
                     onChange={this.handleDateChange}
-                    value={this.state.value}
+                    value={value}
                 />
                 <TimePicker
                     {...this.props.timePickerProps}
                     onChange={this.handleTimeChange}
-                    value={this.state.value}
+                    value={value}
                 />
             </div>
         );
     }
 
     public componentWillReceiveProps(nextProps: IDatePickerProps) {
         if (nextProps.value != null) {
-            this.setState({ value: nextProps.value });
+            this.setState({
+                dateValue: nextProps.value,
+                timeValue: nextProps.value,
+            });
         }
     }
 
     public handleDateChange = (date: Date, isUserChange: boolean) => {
-        const value = DateUtils.getDateTime(date, this.state.value);
+        const value = DateUtils.getDateTime(date, this.state.timeValue);
```

Comment:
merge the objects only for invoking `onChange`. store `date` arg directly in state.

cuz `dateValue` doesn't care at all about its time components and vice-versa for `timeValue` so why bother merging them?